### ISSUE TITLE
AttributeError: 'NoneType' object has no attribute 'group'

### ISSUE THREAD

**Issue by daniellembecker** [gpertea/stringtie#317]
Hello, I cannot seem to work through this error I am getting for line 158 in the prepDE.py script when I use the sample_list.txt file I have made. I have triple checked that the names referenced are correct. Any help would be much appreciated.

Traceback (most recent call last):
  File "prepDE.py", line 158, in <module>
    t_id=RE_TRANSCRIPT_ID.search(v[8]).group(1)
AttributeError: 'NoneType' object has no attribute 'group'

**Comment by gpertea** [comment]
It looks like a transcript ID could not be parsed from the 9th column in a line of a GTF file in that list. Perhaps that GTF file is corrupted somehow?
As a quick way to investigate this, you could edit the script as follows, going to line 158:
* move the `tid=RE_TRANSCRIPT_ID.search(...` line below the `try:` line 
* make sure the `tid=RE_TRANSCRIPT_ID.search(...` is now indented in the same way with the `g_id=getGeneID(...` line below it

After these edits the code at line 158 should look like this (make sure the `try:` line has kept the same indentation with the `except:` line):

```python
  try:
    t_id=RE_TRANSCRIPT_ID.search(v[8]).group(1)
    g_id=getGeneID(v[8], v[0], t_id)
  except:
```
Now when you run the script again you should see the file name and the problematic line printed along with the error message: `"Problem parsing file ..."`.

**Comment by daniellembecker** [comment]
Thank you, I think that my GTF file was corrupted so I will have to work through that. Appreciate the quick response!

**Comment by daniellembecker** [comment]
We are still trying to work through this issue, but I wanted to share what we are noticing thus far in case you have any insight. We are using this genome assembly for our reference .gff3 file:  http://pver.reefgenomics.org/download/Pver_genome_assembly_v1.0.gff3.gz 

We have noticed that after using StringTie to create our .gtf files we see that somehow there is a tab between the 1 and 3 and a true within the parentheses which we think is causing the issue with the transcript ID when trying to loop in the prepDE.py script. Just wondering if you think this is a problem with our source file or your parser?

Example:

C17_R1_001.fastq.gz.sam.sorted.bam.merge.gtf, line 5925:
Pver_Sc0000004_size1560107  StringTie       transcript      1538    1660    1000    -       .       gene_id "STRG.507"; transcript_id "Pver_g535.t**1 3**_prime_partial **true**"; cov "0.000013"; FPKM "0.000004"; TPM "0.000011";

**Comment by daniellembecker** [comment]
Just to update: we were able to use the prepDE.py script once we removed the corruption errors from our reference genome assembly file and re-assemble and estimate the reads to create the gene count matrix file.

**Comment by gpertea** [comment]
I see what was the problem in the reference gff3 file you used -- there were many CDS features like this one (scroll all the way to the right to see the end of the line):
```
Pver_xfSc0000472_size9580	.	CDS	8030	8083	.	-	0	ID=cds.Pver_g23787.t1;Parent=Pver_g23787.t1	5_prime_partial true
```
The problem is that the string `5_prime_partial true` follows the Parent ID there (`Pver_g23787.t1`) separated by a **tab character**, even though I suppose it was meant to be another attribute for that CDS feature, so probably the line should have been formatted like this:
```
Pver_xfSc0000472_size9580	.	CDS	8030	8083	.	-	0	ID=cds.Pver_g23787.t1;Parent=Pver_g23787.t1;5_prime_partial=true
```
Frankly I am a bit embarrassed that my parser simply glossed over that extra tab in the attribute column, because that's not allowed by the GFF3 specification and that violation should've been signaled. The naive assumption was that the 9th columns only has _attribute_=_value_ entries separated by `;` and the parser splits every line in at most 9 columns.. Even though theoretically spaces (but not tabs!) are allowed in the GFF3 specification for the attribute values, using them within ID strings would go against common sense, I'd say.

**Comment by daniellembecker** [comment]
Thank you, yes we removed the 3_prime_partial and 5_prime_partial labels from our original .gff3 file to solve this issue. No worries, I am glad we could find the issue on both ends and now if anyone runs into it, hopefully this post will help!

**Comment by gpertea** [comment]
Ah, only after I answered I looked at the issue you referenced. The fix you suggested there seems to discard those "partialness" info attributes. If you want to keep those as GFF attributes instead of deleting them, I guess a perl expression like this should accomplish that (similar to your `sed` solution, but I don't know how to do that with sed):
```
 perl -pe 's/(Parent=\S+)\s(\S+)\s(\S+)/$1;$2=$3/;s/(Parent=\S+)\s(\S+)\s(\S+)/$1;$2=$3/' \
 Pver_genome_assembly_v1.0.gff3 > fixed.gff3
```
I applied the same regex twice because there are cases where both 5' and 3' partialness info is provided for the same CDS, e.g.
```
Pver_xfSc0007555_size1079	.	CDS	3	1079	.	+	0	ID=cds.Pver_g25526.t1;Parent=Pver_g25526.t1	5_prime_partial true	3_prime_partial true
```
However, `gffread -E` shows other unrelated potential issues with the way that gff3 file was generated -- for example in some cases novel transcript IDs were generated to be the same with the gene IDs, e.g.
```
Pver_Sc0000112_size552208	.	gene	366896	370225	.	+	.	ID=Pver_gene_novel_gene_249_5de57afd;Name=NO_NAME_ASSIGNED
Pver_Sc0000112_size552208	.	mRNA	366896	370225	.	+	.	ID=Pver_gene_novel_model_252_5de57afd;Parent=Pver_gene_novel_gene_249_5de57afd;Name=NO_NAME_ASSIGNED
Pver_Sc0000112_size552208	.	five_prime_UTR	366896	367108	.	+	.	ID=Pver_novel_model_252_5de57afd.utr5p1;Parent=Pver_novel_model_252_5de57afd
```

In this example the mRNA feature has the ID `Pver_gene_novel_model_252_5de57afd` even though its sub-features refer to it as `Pver_novel_model_252_5de57afd` (without the `_gene` part), so the parser will not be able to assign the exon/CDS features to that parent transcript and, what's worse, that transcript will end up being assigned a single "default" exon which misleadingly covers the whole length of the transcript (366896-370225 in this example)! This is quite misleading for StringTie or any other programs trying to make proper use of those novel transcript structures.. So I do believe that the gff3 file should be fixed properly (generated again?) and perhaps analyses like StringTie that used the transcript structures provided there as "guides" etc. should be re-run.
